### ISSUE TITLE
Custom element with custom name

### ISSUE THREAD

**Issue by bernatvadell** [aurelia/aurelia#810]
# Issue
When I use the "customElement" decorator and assign it a custom name, I am not able to load the item.

## 🤔 Expected Behavior
Use a custom element with custom name

## 😯 Current Behavior
When I use a custom element with the naming convention it works great. But when I want to use a custom name, neither with the custom name nor with convection,not load...

## 💻 Code Sample

demo.html
```html
<div>demo text</div>
```

demo.ts
```ts
import { customElement } from "aurelia";

@customElement('demo-custom')
export class Demo {
}
```

my-app.html
```html
<require from="demo"></require>
<div class="message">${message}</div>
<demo-custom></demo-custom>
```

## 🌍 Your Environment

| Software         | Version(s) |
| ---------------- | ---------- |
| Aurelia          | dev
| Language         | Spanish
| Browser          | Edge Chromium
| Bundler          | Webpack
| Operating System | Windows
| NPM/Node/Yarn    | Yarn

**Comment by fkleuver** [comment]
Thanks for reporting. Could you try with the non-convention syntax just to rule out the underlying issue?
```ts
import { customElement } from 'aurelia';
import * as template from './demo.html

@customElement({ name: 'demo-custom', template })
export class Demo {
}
```

**Comment by bernatvadell** [comment]
Hi @fkleuver 

thanks for your fast reply! :)

Now, im receiving this exception when render:

template-element-factory.js:48 Uncaught TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
    at HTMLTemplateElementFactory.createTemplate (template-element-factory.js:48)
    at TemplateCompiler.compile (template-compiler.js:58)
    at RenderContext.compile (render-context.js:109)
    at Controller.hydrateCustomElement (controller.js:150)
    at Function.forCustomElement (controller.js:90)
    at CustomElementRenderer.render (renderer.js:171)
    at Renderer.renderInstructions (renderer.js:91)
    at Renderer.render (renderer.js:68)
    at RenderContext.render (render-context.js:182)
    at Controller.hydrateCustomElement (controller.js:162)

Ty!

**Comment by fkleuver** [comment]
Hm, could you try clearing your node_modules and reinstalling with npm please? We have had issues reported with yarn before, and we are not testing with yarn just yet.

**Comment by bernatvadell** [comment]
Claned and reinstalled with npm install & run with npm start i've same issue.

Im printing with console.log template value, this is the console output:

![image](https://user-images.githubusercontent.com/17298604/72897935-59ebd980-3d23-11ea-964e-3bdec5ccd9e6.png)

Ty

**Comment by fkleuver** [comment]
Ah thanks for clarifying. Odd that the star import doesn't work but the other one does.
The conventions are a custom transformation done by the webpack plugin, something could be going wrong there. @3cp do you know if we have tests for this sort of scenario? Might need to improve that a bit, I can help of course

**Comment by 3cp** [comment]
It is `import template from './demo.html';`. This is expected.

`import * as template` will give you an object, instead of string. Because conventions compiled (preprocessed) html file into a js module. The html files go through our conventions webpack plugin, instead of normal webpack html pipeline.

The ts definition explains the structure, it gives the shape of what customElement decorator wanted. That's how the normal conventions worked.

We aliased the default export to the template value, exactly designed for this manual scenario. So when you get the default export, it gives you the string value of template. But note if your html template contains meta tags (like `<import from="./demo.scss">`), those tags are stripped by conventions processing, **they are not in the default export**. You would need more imports if your html file introduces some dependencies:

```js
import template, {dependencies} from './demo.html';
@customElement({name: 'foo', template, dependencies})
```

https://github.com/aurelia/new/blob/master/app-min/src/resource.d.ts__if_typescript

```ts
declare module '*.html' {
  import { IContainer } from '@aurelia/kernel';
  import { IBindableDescription } from '@aurelia/runtime';
  export const name: string;
  export const template: string;
  export default template;
  export const dependencies: string[];
  export const containerless: boolean | undefined;
  export const bindables: Record<string, IBindableDescription>;
  export const shadowOptions: { mode: 'open' | 'closed'} | undefined;
  export function register(container: IContainer);
}
```

**Comment by fkleuver** [comment]
Thanks @3cp 

What about this then?

```ts
@customElement('demo-custom')
export class Demo {
}
```

Shouldn't that just work?

**Comment by 3cp** [comment]
When class is decorated with certain decorator (`@customElement` is one of them), conventions will skip the class because it thinks user is manually controlling the behaviour.

If `@customElement('demo-custom')` requires further preprocessing, we have to add that to conventions implementation. My guess is `@customElement('demo-custom')` cannot work as it is right now, because au2 has no loader abstraction to **find the paired template**.

We either need add some preprocessing, or user has to opt to more verbose decoration.

**Comment by 3cp** [comment]
@EisenbergEffect what do you think for conventions to re-write `@customElement('demo-custom')`?

**Comment by 3cp** [comment]
@fkleuver I am not sure about the ts type definitions. As `@customElement('demo-custom')` doesn't accept pure string input, tsc will complain at user's code.

Only after conventions re-write it to `@customElement({...viewDef, name: 'demo-custom'})`, it's then following typing definition.

Do you want to create a dummy customElement with `(name: string)` with no implementation to just keep tsc happy at user's code (pre preprocessing code)?

**Comment by bigopon** [comment]
@3cp nice q. Feels like a tough Q to me.

**Comment by 3cp** [comment]
Or easier solution is to introduce `<name as="demo-custom">` meta tag in html template. This is much much easier than deal with js code. Only few lines needed.

We can afford this breaking change, to move custom element name from js code to html template.

**Comment by 3cp** [comment]
Or option No.3 for customElement decorator to deal with.
```js
export class Foo {
  static elementName = "demo-custom";
}
```

**Comment by bernatvadell** [comment]
I like Aurelia because it is very simple to implement without great knowledge of the framework thanks to its conventions.  We should somehow maintain that philosophy.  I think that adding a metatag or a static property does not provide the same clarity as the decorator (My opinion).
 In addition to the effort that would involve migrating the current applications that use Aurelia 1. 

Thank you all for your help.

**Comment by bernatvadell** [comment]
I also like the option of inlineTemplate for those cases where it does not require a large html code.  for example, a component that wraps a datepicker in which the html only exists one input and all the bind work is in the model.  I think in those cases it would be confusing to name the custom element with a metatag ...

**Comment by EisenbergEffect** [comment]
Can this issue be resolved post-aot? Didn't we have some conversations about merging conventions with individual overrides? I seem to recall something around that...

**Comment by jwx** [comment]
@EisenbergEffect I've proposed a while back that individual decorators should be added on top of/merged with conventions while the full decorator, `customElement`, should have full control. Might that be what you're thinking about? (Not sure what happened regarding the proposal, though.)

**Comment by 3cp** [comment]
@jwx let's have a formal RFC to track it if not created.

**Comment by EisenbergEffect** [comment]
I think that's what I was remembering, yes. But I also though @fkleuver was wanting to push some of that out post-AoT.

**Comment by fkleuver** [comment]
In my opinion `@customElement('element-name')` should be a specific exception, because it's the only one of its kind. The `name` of a resource is a special property in several ways:
- It's the only property common across all resource types (ce's, ca's, vc's, bb's, bc's, etc) that is *always required* in the decorator, be it in the object literal or when it's the string-only overload signature.
- It's the only property name not suitable for specifying as a static property as-is, because classes/functions natively also have a `name`.

Based on these two points, I have made the preliminary decision to exclude this (and *only* this) property from our resource convention of allowing each property to be specified as either a standalone decorator or a static property, instead of directly in the resource decorator.

It's not a decision I feel strongly about, and I'll be more than happy to hear counter-arguments. In any case, here's what that means in effect:

### verbose, "native" syntax (not using conventions)
```ts
@customElement({ name: 'some-el', template: '<div>${msg}</div>', containerless: true })
export class SomeEl {}
```

### property-per-decorator
This exists primarily so conventions can be applied more easily, as well as for general consistency. The idea is, it's easier to add a decorator than it is to change `@customElement('foo')` to `@customElement({ name: 'foo', otherStuff })`
```ts
@customElement('some-el')
@template('<div>${msg}</div>')
@containerless
export class SomeEl {}
```

### property-per-static-property
Like property per decorator, this exists for consistency as well as for a declarative syntax that allows you to omit decorators entirely, if you wanted to.
```ts
@customElement('some-el')
export class SomeEl {
  static template = '<div>${msg}</div>';
  static containerless = true;
}
```

### Not possible: `@name`
As per the decision to exclude `name` from this "standard", this would *not* be possible
```ts
@name('some-el')
@template('<div>${msg}</div>')
@containerless
export class SomeEl {}
```

### Not possible: `static name =`
Likewise, no static name property. So this would also *not* be possible:
```ts
@customElement('some-el')
export class SomeEl {
  static name = 'some-el';
  static template = '<div>${msg}</div>';
  static containerless = true;
}
```

### So how to specify *just* the `name` then?

Via the `@customElement` decorator. It's the only overload that every resource decorator shares in common that accepts just one of the properties, because it's the only required property:
```ts
@customElement('some-el')
export class SomeEl {}
```

@3cp 
To sum it up, I would say treat `@customElement('some-el')` as if it were `@name('some-el')`, because that is the role it plays in the api. The only difference with the other decorators (e.g. `@useShadowDOM`) is that this one needs to be removed or rewritten. You'd just look for a `Decorator` with an `Expression` with `SyntaxKind.StringLiteral` (as far as I'm concerned as soon as someone puts a string variable in there, we're going into the area of things we do post-AOT, so let's keep it simple like this for the conventions plugin?)

**Comment by 3cp** [comment]
@fkleuver that sounds convincing, I can pickup and rewrite that decorator in conventions. I will work on it early next week. However, we need the type definition to be ready first before merging new conventions implementation, ping me when you added the new typing.

**Comment by EisenbergEffect** [comment]
@fkleuver Thanks for great explanation. I agree on this assessment.

**Comment by HamedFathi** [comment]
@fkleuver 

What is your opinion about `@customAttribute("foo")`?

**Comment by 3cp** [comment]
@HamedFathi that already worked. All customAttribute, valueConverter and bindingBehavior is same as aurelia 1.

Custom element has trouble only because there is a paired html file, `@customElement('foo')` doesn't imply any dependency on paired file.

**Comment by HamedFathi** [comment]
@3cp 

I asked this because if you recall, I mentioned it before in Discord.

> I want to define just a custom name for my component, Why should I define template too?
> I expect to achieve convention over configuration in this scenario too.

![unknown](https://user-images.githubusercontent.com/8418700/73117527-28118780-3f5c-11ea-9098-ec28cd4a4ebe.png)

**Comment by fkleuver** [comment]
> Custom element has trouble only because there is a paired html file, @customElement('foo') doesn't imply any dependency on paired file.

Correct. To further on that, `@customElement` is the only resource decorator that requires more than just the `name` to work. All the other ones would produce working resource descriptions even without conventions.

(nitpick: *technically* customElement works without `template` as well, but in that case you'd explicitly say `@customElement({ name: 'foo', template: null })`. This would be equivalent to `@noView` in v1)

**Comment by fkleuver** [comment]
> However, we need the type definition to be ready first before merging new conventions implementation, ping me when you added the new typing.

Type definitions for what though? The `@customElement` decorator has already had this overload for a while: <https://github.com/aurelia/aurelia/blob/master/packages/runtime/src/resources/custom-element.ts#L118>

**Comment by 3cp** [comment]
Oh, I see. I am surprised it's there. I guess it's just a placeholder without implementation?

**Comment by fkleuver** [comment]
It has been implemented for a while. The runtime behavior of that overload without conventions is a `null` template.
In what situation is TS giving you errors anyway? Wouldn't make sense if `@customElement('foo')` gave you errors, people (including the person who opened this issue) have been doing this for months. 🤔

**Comment by 3cp** [comment]
No. I was just speculating error. Btw, since the current api behavior is null template, would that be a "breaking change" if conventions changed its meaning?

**Comment by fkleuver** [comment]
Does conventions do anything if there is no matching html file?

**Comment by 3cp** [comment]
Nope, I forgot that. Perfect.

**Comment by 3cp** [comment]
Confirmed working with latest release.
